ledger-tool/src/main.rs

```diff
@@ -1332,7 +1335,27 @@ fn main() {
                         .takes_value(false)
                         .help("Do not print account data when printing account contents."),
                 )
-                .arg(&max_genesis_archive_unpacked_size_arg),
+                .arg(
+                    Arg::with_name("account")
+                        .long("account")
+                        .takes_value(true)
+                        .value_name("PUBKEY")
+                        .validator(is_pubkey)
+                        .multiple(true)
+                        .help(
+                            "Limit output to accounts corresponding to the specified pubkey(s), \
+                            may be specified multiple times",
+                        ),
+                )
+                .arg(
+                    Arg::with_name("program_accounts")
```

Comment:
Any specific reason not to make this a `multiple(true)` argument, just as the `--account`?
I can think of, for example, querying all the deployed programs using the loader program as a reference.  And, IIUC, we have 4 or 5 loader program addresses, right?

Reply:
`Bank::get_program_accounts()` only accepts a single pubkey. That being said, we could use `Bank::scan_all_accounts()` and just do the filtering ourselves if we wanted to allow searching for multiple keys.

If we did that, it would be nice if we could have an extra layer of organization in the json, ie:
`{
    "key1": [
        ...
    ],
    "key2": [
        ...
    ],
    ...
}
`
but with the way we stream them, that wouldn't be possible. Outputting them in separate "sections" would require accumulating all of them and then writing at the end

Reply:
There is a loop for `pubkeys` here: https://github.com/solana-labs/solana/pull/34925/files#diff-15618f2c21fa1d13eb4bfc3b430f552cb85831801d1a742f68505c57ac54d2ebR709-R710

Why not just loop similarly for the program accounts?
Is there a restriction on the order that we want the accounts to be output in?

Or are you saying that we would not see which account belongs to which program?
I wonder if this is too much work to extend [`output_account()`](https://github.com/steviez/solana/blob/26d62b9516b4c079f17f662a03874d85c0719176/ledger-tool/src/main.rs#L105) to print this information.
Though, maybe it is not something we always want to print.

Reply:
If you go look up the impl of `Bank::get_program_accounts()`, it scans the entire accountsdb. Yesterday on mnb, this took about 1 hour to complete. The loop you referenced is looking up a single account, and is pretty fast (because we know exactly what we're looking for)

While what you described (looping over pubkeys and calling `Bank::get_program_accounts()` for each) would work / allow the "separation" I described, it would also make the runtime scale linearly. Ideally, we could complete everything on one pass

Reply:
I didn't know `get_program_accounts()` is this slow.
Maybe add a note to the argument description?
I assume there is no progress indicator for this scan progress?

It seems relatively easy to extend `Bank::get_program_accounts() -> AccountsDb::load_by_program()` to accept a list, but it is probably outside the scope you are targeting here.

In any case, when there is no way to scan more efficiently, multiple `--program-account` arguments is no slower than running the whole tool multiple times.

But I also do not want to stall this PR due to this discussion.

Reply:
> It seems relatively easy to extend Bank::get_program_accounts() -> AccountsDb::load_by_program() to accept a list

Yep, I had a similar suggestion higher up in the thread that is essentially equivalent:

_Bank::get_program_accounts() only accepts a single pubkey. That being said, we could use Bank::scan_all_accounts() and just do the filtering ourselves if we wanted to allow searching for multiple keys._

> but it is probably outside the scope you are targeting here.

Ha yup

> In any case, when there is no way to scan more efficiently, multiple --program-account arguments is no slower than running the whole tool multiple times.

Yep, completely agree.

I'd be open to extending to allow scanning for multiple "program_accounts" simultaneously, but to agree with your earlier point, let's just get this in now